ramips: add Teltonika RUT9xx series support

simsasss opened the pull request:
This pull request add support for RUT9xx series.
The Teltonika RUT9XX series is an industrial mobile router series, it supports WiFi 5, have CAT 4 modem and various I/Os like RS232, RS485.
Base boards are the same just modems and I/O differ depending on model.

Specifications:
- MediaTek MT7628AN SoC
- 128 MB of RAM EtronTech EM68C16CWQG-25IH
- 16MB of Flash Winbond W25Q128 SPI
- 4x 10/100 Mbps Ethernet, with passive PoE support on LAN1
- MediaTek MT7628AN 2.4 GHz 802.11n WiFi
- cat 4 modem
- RS232, DB9 connector, Cypress ACM via USB, /dev/ttyACM0
- RS485 /dev/ttyS1
- microSD card slot
- 2.0 USB Type-A HOST port
- analog 0-24V input (MCP3221)
- Relay
- 2x Digital input
- 2x Digital output
- 2x SIM slot (can be swapped via GPIO)
- eSIM (depends on hardware configuration, can be swapped via GPIO)

Flashing via OEM WebUI:
1. Download the firmware image *-squashfs-factory.bin
2. Upload firmware image via OEM WebUI firmware update, do not keep settings

To revert back to OEM firmware:
https://wiki.teltonika-networks.com/view/Bootloader_menu

simsasss committed:
ramips: mt76x8: add Teltonika RUT956/RUT906 support

Specification:
- MediaTek MT7628AN SoC
- 128 MB of RAM EtronTech EM68C16CWQG-25IH
- 16MB of Flash Winbond W25Q128 SPI
- 4x 10/100 Mbps Ethernet, with passive PoE support on LAN1
- MediaTek MT7628AN 2.4 GHz 802.11n WiFi
- Quectel EC25-EUX (RUT956) or Meig SLM770A (RUT906) modem
- GNSS
- RS232, DB9 connector, Cypress ACM via USB, /dev/ttyACM0
- RS485 /dev/ttyS1
- microSD card slot
- 2.0 USB Type-A HOST port
- analog 0-24V input (MCP3221)
- Relay
- 2x Digital input
- 2x Digital output
- 2x SIM slot (can be swapped via GPIO)
- eSIM (depends on hardware configuration, can be swapped via GPIO)

GPIO:
 - 1 button (Reset)
 - 7 LEDs (power, status green/red, RSSI 1,2,3,4,5)
 - 5 Modem control (power button, reset, status (v5 hw), SIM select, eSIM select)
 - 2 Digital input
 - 2 Digital output
 - 1 Isolated input
 - 1 RS485 tx enable
 - 1 RS485 rx enable
 - 1 Relay
 - 1 Data Carrier Detect (DCD)

Flashing via OEM WebUI:
1. Download the firmware image *-squashfs-factory.bin
2. Upload firmware image via OEM WebUI firmware update, do not keep settings

To revert back to OEM firmware:
https://wiki.teltonika-networks.com/view/Bootloader_menu

Mobile data connection:
Meig SLM770 create proto DHCP interface with usb0 device.
Quectel EC25-EUX create proto qmi interface.

Signed-off-by: Simonas Tamošaitis <simsasss@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18788
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

musashino205 commented on target/linux/ramips/dts/mt7628an_teltonika_rut9x6.dtsi:
```diff
@@ -0,0 +1,76 @@
+
+&i2c {
+	status = "okay";
+	hwmon@0x4d {
```
The node name shouldn't have the prefix `0x` in the hexadecimal unit address.

```suggestion
	hwmon@4d {
```

musashino205 commented on target/linux/ramips/dts/mt7628an_teltonika_rut9x6.dtsi:
```diff
@@ -0,0 +1,76 @@
+		reference-voltage-microvolt = <3300000>;
+	};
+
+	hwmon@0x48 {
```
ditto

musashino205 commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -179,6 +179,13 @@ ramips_setup_interfaces()
+	teltonika,rut956)
+		ucidef_add_switch "switch0" \
+			"0:lan" "1:lan" "2:lan" "4:wan" "6t@eth0"
+	;;
```
```suggestion
		;;
```

hauke commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -636,6 +638,61 @@ define Device/tama_w06
+endef
+TARGET_DEVICES += teltonika_rut906
+
+define Device/teltonika_rut951
+  DEVICE_VENDOR := Teltonika
+  DEVICE_MODEL := RUT951
+  IMAGE_SIZE := 15424k
+  BLOCKSIZE := 64k
+  DEVICE_PACKAGES += uqmi kmod-mt76x2 kmod-usb2 kmod-usb-ohci kmod-usb-serial-option kmod-spi-gpio kmod-gpio-nxp-74hc164 kmod-i2c-mt7628
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | check-size | append-teltonika-metadata rut9m \
+  "mod1":"2c7c_6005" "mod2":"TLA2021" "mod3":"CH343" "mod4":"esim" "mod5":"ala440"
+  IMAGE/sysupgrade.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | check-size | append-metadata
+endef
+TARGET_DEVICES += teltonika_rut951
```
The RUT951 and the RUT901 are using the same device tree except the name. The device definition is also mostly the same, it just has different device packages.

I would prefer to only add one device definition and use it for both boards. 

Same for the RUT906 and RUT956.

simsasss replied:
Hello @hauke,
could you refer to any example?
Do you mean I should do like "7links_wlr-1230" is done?
```
define Device/7links_wlr-1230
  $(Device/7links_wlr-12xx)
  DEVICE_MODEL := WLR-1230
endef
TARGET_DEVICES += 7links_wlr-1230
```
Thank you.

hauke replied:
I would prefer if you remove RUT951 entry and dts file and then add this to the RUT901:
```
DEVICE_ALT0_VENDOR := Teltonika
DEVICE_ALT0_MODEL := RUT951
```
maybe change the `model` in the device tree to contain both names.

Then we would only build one image which can be loaded on both boards. We would have less images. 

I assume that the devices do not need a different OpenWrt image, is that correct?

simsasss committed:
ramips: mt76x8: add Teltonika RUT951/RUT901 support

Specification:
- MediaTek MT7628AN SoC
- 128 MB of RAM EtronTech EM68C16CWQG-25IH
- 16MB of Flash Winbond W25Q128 SPI
- 4x 10/100 Mbps Ethernet, with passive PoE support on LAN1
- MediaTek MT7628AN 2.4 GHz 802.11n WiFi
- Quectel EC25-EUX 4G cat 4 modem (RUT951) or Quectel EC200A-EU 4G cat 4 modem (RUT901)
- 1x Digital input
- 1x Digital output
- 2x SIM slot (can be swapped via GPIO)
- eSIM (depends on hardware configuration, can be swapped via GPIO)

GPIO:
 - 1 button (Reset)
 - 7 LEDs (power, status green/red, RSSI 1,2,3,4,5)
 - 5 Modem control (power button, reset, status, SIM select, eSIM select)
 - 1 Digital input
 - 1 Digital output

Flashing via OEM WebUI:
1. Download the firmware image *-squashfs-factory.bin
2. Upload firmware image via OEM WebUI firmware update, do not keep settings

To revert back to OEM firmware:
https://wiki.teltonika-networks.com/view/Bootloader_menu

Mobile data connection:
Quectel EC200A-EU send command to modem:
    echo -ne 'AT+QNETDEVCTL=3,1\r\n' > /dev/ttyUSB2
    Create DHCP interface with usb0 device.
Quectel EC25-EUX create proto qmi interface.

Signed-off-by: Simonas Tamošaitis <simsasss@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18788
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
